Autogen config docs

jam10o-new opened the pull request:
Somewhat "working", might not be the ideal format, but can change anything if needed.
part of: https://github.com/paritytech/parity-ethereum/pull/10067

Commit:
skeleton of changes

Commit:
config docs first attempt

axelchalon commented:
Fine by me if it does the job! :+1:
cc @tomusdrw 

From an architecture POV it might make sense in the future to put `generate-config-data.js` (or the gist of it) in its own/separate repo. Then, we could use it as dependency in Parity Config Generator and have `generate-config-doc.js` be in the paritytech wiki repo and executed there, which would make more sense.

jam10o-new commented:
Should I merge this in that case, or do we still need to do reviews @axelchalon?

axelchalon commented:
I'd wait for Tomek to have a look

tomusdrw commented on scripts/generate-config-docs.js:
```diff
@@ -0,0 +1,118 @@
+
+const output_dir = "./docs/"
+const outpit_file = "config.md"
+const pre_example_config = new BufferList("---\ntitle: Configuring Parity Ethereum\n---\n\nParity can be configured using either the [CLI options](#cli-options) or a [config file](#config-file). Should the CLI flags and the config file disagree about a setting, the CLI takes precedence.\n\nYou can list all CLI options by running `$parity --help`. The vast majority of CLI options map to a setting  ...
```
maybe use \` (backtick) and format the newlines properly? (although you would need to escape inline backticks).

I think it would be best to put this to a separate file and just `fs.readFileSync` from the template file.

jam10o-new replied:
made a separate file with all the template stuff but I'm not sure if it's what you were thinking of :sweat_smile:

tomusdrw commented on scripts/generate-config-docs.js:
```diff
@@ -0,0 +1,118 @@
+const pre_config_doc = new BufferList("## Presets \nParity can also be launched with a [preset configuration file](https://github.com/paritytech/parity-ethereum/tree/master/parity/cli/presets) using the `--config` flag with one of the following values:\n- `dev`: uses [dev chain specifications](Private-development-chain) with [Instant-seal](Pluggable-Consensus#instant-seal) consensus engine. The g ...
+const post_config_doc = new BufferList("");
+
+function dirtySerialize (valueItem) {
```
please correct the indentation, actually could you run the entire file through `semistandard/standard` (whatever is used here) and make sure the syntax is checked on every build?

tomusdrw commented on scripts/generate-config-data.js:
```diff
@@ -289,3 +289,8 @@ function augment (data, extra) {
   console.error(e);
   process.exit(1);
 });
+
```
Seems that this file now:
1. Runs a script to fetch some stuff.
2. But also exports some functions.

I think it would be best to distinguish if it's imported or run directly to prevent running when importing as a module.
AFAIR you can just check
```js
if (!module.parent) {
 // run stuff
} else {
  module.exports = {
   fetchSource,
   getCliOptions
  };
}

```

jam10o-new replied:
did exactly this - also, made sure that we're only fetching once in this entire process

tomusdrw commented on docs/config.md:
```diff
@@ -0,0 +1,2448 @@
+---
```
Is that required / wanted in this repo? Seems it's quite big and might get out of date anyway, maybe we can generate it dynamically (+ add to .gitignore) and just push to `gh-pages` instead?

jam10o-new replied:
added to gitignore :)

tomusdrw reviewed (commented):
Seems pretty cool! I'd clean up the code a bit though and will be glad to merge.

Commit:
grumbles
use generate-config-data as a library, do not fetch twice
if running within the publish-docs script, should grab local mod.rs (not tested yet)
seperate docTemplate file (maybe not using templates properly)
linting and formatting

jam10o-new commented:
I need to test whether the change I made to generate-config-data (grab local mod.rs if it exists) actually works, but I'll ask y'all to look at this again if everything works properly.

edit: this doesn't seem to work yet :(

jam10o-new commented:
was not an issue on the js side- fixed in commit https://github.com/paritytech/parity-ethereum/pull/10067/commits/169c30a8c25fee57e635be170d34990f90422637

tomusdrw commented on scripts/generate-config-data.js:
```diff
@@ -13,19 +13,23 @@ const CONFIG_IS = {
+        }));
+      });
+    } else {
+      resolve(fs.readFileSync(path.resolve(__dirname, '../../parity/cli/mod.rs'), 'UTF-8'));
```
Maybe this:
```js
const util = require('util');
...
function fetchSource() {
  if (process.env.AUTOGENSCRIPT) {
     return util.promisify(fs.readFile)(path.resolve(__dirname, '../../parity/cli/mod.rs'), 'UTF-8'));
  }
  
  return https.get(...)
     ....
}

jam10o-new replied:
readFile is already an async function, so it returns a promise :+1:

tomusdrw commented on docs/docTemplate.js:
```diff
@@ -0,0 +1,49 @@
+const BufferList = require('bl');
+
+module.exports = {
+  preExampleConfig: new BufferList(`
```
I was thinking of storing this templates in a regular `.md` files (since you would get syntax highlighting) and then load them via `preExampleConfig: fs.readFileSync('./template/pre-example.md', 'utf8')`. But this is good enough as well :)

tomusdrw reviewed (approved):
Looks good!

Commit:
quick readability fix

jam10o-new commented:
ack - shouldn't have merged this before making sure it's run from the right working dir, will make another pr for future changes
